### ISSUE TITLE
[Bug]: py_pytest_main triggers assertion error in coverage

### ISSUE THREAD

**Issue by garymm** [aspect-build/rules_py#525]
### What happened?

I wrote a macro to add a pytest main with coverages support based on https://github.com/aspect-build/rules_py/pull/492.

It seems to trigger an assertion error in the coverage library. Not sure if this is a bug in the coverage library or not. I found this similar failure which suggests something to do with concurrency: 
https://github.com/nedbat/coveragepy/issues/488
https://github.com/nedbat/coveragepy/issues/1506

So maybe this is a bug in coverage, but it suggests py_pytest_main is doing something unusual that is triggering?

Test output:

```
Executing tests from //earl/environment_loop:test_common
-----------------------------------------------------------------------------
============================= test session starts ==============================
platform linux -- Python 3.12.8, pytest-8.3.4, pluggy-1.5.0 -- /home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/bin/earl/environment_loop/test_common.runfiles/.test_common.venv/bin/python3
rootdir: /home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/bin/earl/environment_loop/test_common.runfiles/_main
plugins: jaxtyping-0.2.37
collecting ... collected 4 items

earl/environment_loop/test_common.py::TestCommon::test_multi_observe_cycle PASSED [ 25%]
earl/environment_loop/test_common.py::TestCommon::test_pixel_obs_to_video_observe_cycle_invalid_shape PASSED [ 50%]
earl/environment_loop/test_common.py::TestCommon::test_pixel_obs_to_video_observe_cycle_multiple_videos PASSED [ 75%]
earl/environment_loop/test_common.py::TestCommon::test_pixel_obs_to_video_observe_cycle_single_video PASSED [100%]

- generated xml file: /home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/testlogs/earl/environment_loop/test_common/test.xml -
============================== 4 passed in 8.96s ===============================
self._collectors:
  <Collector at 0x77d4e0662f90: CTracer>
                      <module> : /home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/bin/earl/environment_loop/test_common.runfiles/_main/earl/environment_loop/test_common_pytest_main.py:32
                         start : /home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/bin/earl/environment_loop/test_common.runfiles/.test_common.venv/lib/python3.12/site-packages/coverage/control.py:640
               _init_for_start : /home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/bin/earl/environment_loop/test_common.runfiles/.test_common.venv/lib/python3.12/site-packages/coverage/control.py:540
                      __init__ : /home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/bin/earl/environment_loop/test_common.runfiles/.test_common.venv/lib/python3.12/site-packages/coverage/collector.py:125
  <Collector at 0x77d4495636e0: CTracer>
                      <module> : /home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/bin/earl/environment_loop/test_common.runfiles/_main/earl/environment_loop/test_common_pytest_main.py:88
                          main : /home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/bin/earl/environment_loop/test_common.runfiles/.test_common.venv/lib/python3.12/site-packages/_pytest/config/__init__.py:175
                      __call__ : /home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/bin/earl/environment_loop/test_common.runfiles/.test_common.venv/lib/python3.12/site-packages/pluggy/_hooks.py:513
                     _hookexec : /home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/bin/earl/environment_loop/test_common.runfiles/.test_common.venv/lib/python3.12/site-packages/pluggy/_manager.py:120
                    _multicall : /home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/bin/earl/environment_loop/test_common.runfiles/.test_common.venv/lib/python3.12/site-packages/pluggy/_callers.py:103
           pytest_cmdline_main : /home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/bin/earl/environment_loop/test_common.runfiles/.test_common.venv/lib/python3.12/site-packages/_pytest/main.py:330
                  wrap_session : /home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/bin/earl/environment_loop/test_common.runfiles/.test_common.venv/lib/python3.12/site-packages/_pytest/main.py:283
                         _main : /home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/bin/earl/environment_loop/test_common.runfiles/.test_common.venv/lib/python3.12/site-packages/_pytest/main.py:336
                      __call__ : /home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/bin/earl/environment_loop/test_common.runfiles/.test_common.venv/lib/python3.12/site-packages/pluggy/_hooks.py:513
                     _hookexec : /home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/bin/earl/environment_loop/test_common.runfiles/.test_common.venv/lib/python3.12/site-packages/pluggy/_manager.py:120
                    _multicall : /home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/bin/earl/environment_loop/test_common.runfiles/.test_common.venv/lib/python3.12/site-packages/pluggy/_callers.py:103
             pytest_collection : /home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/bin/earl/environment_loop/test_common.runfiles/.test_common.venv/lib/python3.12/site-packages/_pytest/main.py:347
               perform_collect : /home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/bin/earl/environment_loop/test_common.runfiles/.test_common.venv/lib/python3.12/site-packages/_pytest/main.py:809
                      genitems : /home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/bin/earl/environment_loop/test_common.runfiles/.test_common.venv/lib/python3.12/site-packages/_pytest/main.py:975
                      genitems : /home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/bin/earl/environment_loop/test_common.runfiles/.test_common.venv/lib/python3.12/site-packages/_pytest/main.py:975
                      genitems : /home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/bin/earl/environment_loop/test_common.runfiles/.test_common.venv/lib/python3.12/site-packages/_pytest/main.py:975
                      genitems : /home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/bin/earl/environment_loop/test_common.runfiles/.test_common.venv/lib/python3.12/site-packages/_pytest/main.py:970
             _collect_one_node : /home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/bin/earl/environment_loop/test_common.runfiles/.test_common.venv/lib/python3.12/site-packages/_pytest/main.py:835
              collect_one_node : /home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/bin/earl/environment_loop/test_common.runfiles/.test_common.venv/lib/python3.12/site-packages/_pytest/runner.py:567
                      __call__ : /home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/bin/earl/environment_loop/test_common.runfiles/.test_common.venv/lib/python3.12/site-packages/pluggy/_hooks.py:513
                     _hookexec : /home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/bin/earl/environment_loop/test_common.runfiles/.test_common.venv/lib/python3.12/site-packages/pluggy/_manager.py:120
                    _multicall : /home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/bin/earl/environment_loop/test_common.runfiles/.test_common.venv/lib/python3.12/site-packages/pluggy/_callers.py:103
    pytest_make_collect_report : /home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/bin/earl/environment_loop/test_common.runfiles/.test_common.venv/lib/python3.12/site-packages/_pytest/runner.py:391
                     from_call : /home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/bin/earl/environment_loop/test_common.runfiles/.test_common.venv/lib/python3.12/site-packages/_pytest/runner.py:341
                       collect : /home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/bin/earl/environment_loop/test_common.runfiles/.test_common.venv/lib/python3.12/site-packages/_pytest/runner.py:389
                       collect : /home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/bin/earl/environment_loop/test_common.runfiles/.test_common.venv/lib/python3.12/site-packages/_pytest/python.py:549
_register_setup_module_fixture : /home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/bin/earl/environment_loop/test_common.runfiles/.test_common.venv/lib/python3.12/site-packages/_pytest/python.py:562
                           obj : /home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/bin/earl/environment_loop/test_common.runfiles/.test_common.venv/lib/python3.12/site-packages/_pytest/python.py:284
                       _getobj : /home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/bin/earl/environment_loop/test_common.runfiles/.test_common.venv/lib/python3.12/site-packages/_pytest/python.py:546
              importtestmodule : /home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/bin/earl/environment_loop/test_common.runfiles/.test_common.venv/lib/python3.12/site-packages/_pytest/python.py:493
                   import_path : /home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/bin/earl/environment_loop/test_common.runfiles/.test_common.venv/lib/python3.12/site-packages/_pytest/pathlib.py:587
                 import_module : /home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/external/rules_python++python+python_3_12_x86_64-unknown-linux-gnu/lib/python3.12/importlib/__init__.py:90
                   _gcd_import : <frozen importlib._bootstrap>:1387
                _find_and_load : <frozen importlib._bootstrap>:1360
       _find_and_load_unlocked : <frozen importlib._bootstrap>:1331
                _load_unlocked : <frozen importlib._bootstrap>:935
                   exec_module : /home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/bin/earl/environment_loop/test_common.runfiles/.test_common.venv/lib/python3.12/site-packages/_pytest/assertion/rewrite.py:184
                      <module> : /home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/bin/earl/environment_loop/test_common.runfiles/_main/earl/environment_loop/test_common_pytest_main.py:32
                         start : /home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/bin/earl/environment_loop/test_common.runfiles/.test_common.venv/lib/python3.12/site-packages/coverage/control.py:640
               _init_for_start : /home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/bin/earl/environment_loop/test_common.runfiles/.test_common.venv/lib/python3.12/site-packages/coverage/control.py:540
                      __init__ : /home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/bin/earl/environment_loop/test_common.runfiles/.test_common.venv/lib/python3.12/site-packages/coverage/collector.py:125
Traceback (most recent call last):
  File "/home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/bin/earl/environment_loop/test_common.runfiles/_main/earl/environment_loop/test_common_pytest_main.py", line 94, in <module>
    cov.stop()
  File "/home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/bin/earl/environment_loop/test_common.runfiles/.test_common.venv/lib/python3.12/site-packages/coverage/control.py", line 668, in stop
    self._collector.stop()
  File "/home/user/.cache/bazel/_bazel_user/ec846536f2be8e71d7492ef5e09babf0/sandbox/processwrapper-sandbox/82/execroot/_main/bazel-out/k8-fastbuild/bin/earl/environment_loop/test_common.runfiles/.test_common.venv/lib/python3.12/site-packages/coverage/collector.py", line 353, in stop
    assert self._collectors[-1] is self, (
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: Expected current collector to be <Collector at 0x77d4e0662f90: CTracer>, but it's <Collector at 0x77d4495636e0: CTracer>
```

```starlark
"""Wrapper macros for aspect_rules_py py_test and py_pytest_main."""

load("@aspect_rules_py//py:defs.bzl", "py_pytest_main", aspect_py_test = "py_test")

def py_test(name, srcs, main = None, deps = [], **kwargs):
    """A macro that wraps the py_test rule from aspect_rules_py.

    This macro sets a default main attribute that points to the pytest_main target if not provided.

    Args:
      name: A string representing the test name.
      srcs: A list of source file names.
      deps: A list of dependencies.
      main: An optional string for the main target. If not provided, a default pytest_main target is generated.
      **kwargs: Additional keyword arguments passed to the underlying py_test rule.

    Usage:
      load("//tools/py_test:py_test.bzl", "py_test")
      py_test(
          name = "my_test",
          srcs = ["my_test.py"],
          deps = [...],
      )
    """
    if main == None:
        main_name = name + "_pytest_main"
        py_pytest_main(
            name = main_name,
            deps = [
                "@pypi//coverage",
                "@pypi//pytest",
            ],
        )
        main = ":" + main_name + ".py"
        srcs.append(":" + main_name)
        deps.append(":" + main_name)
    aspect_py_test(
        name = name,
        srcs = srcs,
        main = main,
        deps = deps,
        **kwargs
    )
```

### Version

Development (host) and target OS/architectures: Linux x86_64

Output of `bazel --version`: 
bazel 8.0.1

Version of the Aspect rules, or other relevant rules from your
`WORKSPACE` or `MODULE.bazel` file: rules_py 1.3.1

### How to reproduce

```shell
Use the above macro to declare a py_test and run it with `bazel coverage`.
```

### Any other information?

I'd actually prefer not to use pytest at all if coverage were supported without it (https://github.com/aspect-build/rules_py/issues/517)

**Comment by garymm** [comment]
Interestingly, changing the macro to set the name as a dunder seems to fix!

```starlark
        main_name = "__" + name + "_pytest_main__"
```
